drivers/i2c/i2c_cc32xx.c

```diff
@@ -0,0 +1,417 @@
+		/* No errors, now check what we need to do next */
+		switch (data->state) {
+		/*
+		 * ERROR case is OK because if an Error is detected, a STOP bit
+		 * is sent; which in turn will call another interrupt. This
+		 * interrupt call will then post the transfer_complete semaphore
+		 * to unblock the transfer function.
+		 */
+		case I2C_CC32XX_ERROR:
+		case I2C_CC32XX_IDLE_MODE:
+			/* Indicate transfer complete */
+			k_sem_give(&data->transfer_complete);
+			break;
+
+		case I2C_CC32XX_WRITE_MODE:
+			_i2c_cc32xx_isr_handle_write(base, data);
+			break;
+
+		case I2C_CC32XX_READ_MODE:
+			_i2c_cc32xx_isr_handle_read(base, data);
+			break;
+
+		default:
+			data->state = I2C_CC32XX_ERROR;
+			break;
+		}
+	} else {
+		/* Some sort of error happened */
+		data->state = I2C_CC32XX_ERROR;
+
+		if (err_status & (I2C_MASTER_ERR_ARB_LOST |
+				  I2C_MASTER_ERR_ADDR_ACK)) {
+			k_sem_give(&data->transfer_complete);
+		} else {
+			/* Send a STOP bit to end I2C communications */
+			/*
+			 * I2C_MASTER_CMD_BURST_SEND_ERROR_STOP -and-
+			 * I2C_MASTER_CMD_BURST_RECEIVE_ERROR_STOP
+			 * have the same values
+			 */
+			MAP_I2CMasterControl(base,
+					  I2C_MASTER_CMD_BURST_SEND_ERROR_STOP);
+			k_sem_give(&data->transfer_complete);
+		}
+	}
+}
+
+static int i2c_cc32xx_init(struct device *dev)
+{
+	u32_t base = DEV_BASE(dev);
+	const struct i2c_cc32xx_config *config = DEV_CFG(dev);
+	struct i2c_cc32xx_data *data = DEV_DATA(dev);
+	u32_t bitrate_cfg;
+	int error;
+	u32_t regval;
+
+	k_sem_init(&data->mutex, 1, UINT_MAX);
+	k_sem_init(&data->transfer_complete, 0, UINT_MAX);
+
+	/* In case of app restart: disable I2C module, clear NVIC interrupt */
+	/* Note: this was done *during* pinmux setup in SimpleLink SDK. */
+	MAP_I2CMasterDisable(base);
+
+	/* Clear interrupt INT_I2CA0 */
+	MAP_IntPendClear((unsigned long)(config->exception_no + 16));
+
+	configure_i2c_irq(config);
+
+	data->state = I2C_CC32XX_IDLE_MODE;
+
+	/* Take I2C hardware semaphore. */
+	regval = HWREG(COMMON_REG_BASE);
+	regval = (regval & ~I2C_SEM_MASK) | (0x01 << I2C_SEM_TAKE);
+	HWREG(COMMON_REG_BASE) = regval;
+
+	/* Set to default configuration: */
+	bitrate_cfg = _i2c_map_dt_bitrate(config->bitrate);
+	error = i2c_cc32xx_configure(dev, I2C_MODE_MASTER | bitrate_cfg);
+	if (error) {
+		return error;
+	}
+
+	/* Clear any pending interrupts */
+	MAP_I2CMasterIntClear(base);
+
+	/* Enable the I2C Master for operation */
+	MAP_I2CMasterEnable(base);
+
+	/* Unmask I2C interrupts */
+	MAP_I2CMasterIntEnable(base);
+
+	return 0;
+}
+
+static const struct i2c_driver_api i2c_cc32xx_driver_api = {
+	.configure = i2c_cc32xx_configure,
+	.transfer = i2c_cc32xx_transfer,
+};
+
+
+static const struct i2c_cc32xx_config i2c_cc32xx_config = {
+	.base = CONFIG_I2C_0_BASE_ADDRESS,
+	.bitrate = CONFIG_I2C_0_BITRATE,
+	.exception_no = CONFIG_I2C_0_IRQ,
+};
+
+static struct i2c_cc32xx_data i2c_cc32xx_data;
+
+DEVICE_AND_API_INIT(i2c_cc32xx, TI_CC32XX_I2C_40020000_LABEL, &i2c_cc32xx_init,
```

Comment:
TI_CC32XX_I2C_40020000_LABEL should also go through fixup.

Reply:
Yup, missed that one.